most-used-words.el

```diff
@@ -50,97 +50,170 @@
 
+(defgroup most-used-words nil
+  "Display most used words in buffer."
+  :prefix "most-used-words-"
+  :group 'tool
+  :link '(url-link :tag "Repository" "https://github.com/udyantw/most-used-words"))
+
+(defcustom most-used-words-word-display 3
+  "Default word data to display."
+  :type 'integer
+  :group 'most-used-words)
+
+(defcustom most-used-words-display-type 'buffer
+  "The method to display most used words."
+  :type '(choice (const :tag "buffer" buffer)
+                 (const :tag "table" table))
+  :group 'most-used-words)
+
+(defvar most-used-words--buffer-name "*Most used words*"
+  "Buffer name for displaying most used words.")
+
+(defvar most-used-words--buffer nil
+  "Recorded buffer for currently doing most used words.")
+
+;;; Util
+
 (cl-defmacro most-used-words-with-view-buffer (buffer &body body)
+  "Default BUFFER for this package to display and execute BODY."
   `(progn
-     (with-current-buffer ,buffer
-       ,@body)
-     (view-buffer-other-window ,buffer
-			       nil
-			       #'kill-buffer)))
+     (with-current-buffer ,buffer ,@body)
+     (view-buffer-other-window ,buffer nil #'kill-buffer)))
+
+(defun most-used-words--form-data (word count percent)
+  "Form the plist data item with WORD, COUNT, PERCENT."
+  (list :word word :count count :percent percent))
+
+(defun most-used-words-data (n)
+  "Form used word data base on N count of data displayed."
+  (let* ((most-used (most-used-words-buffer-1 n t))
+         (word-counts (cl-first most-used))
+         (total-count (float (cl-second most-used)))
+         (data-lst '())  data-item)
+    (cl-loop for (word count) in word-counts
+             do
+             (setq data-item (most-used-words--form-data word count (* 100 (/ count total-count))))
+             (push data-item data-lst))
+    (reverse data-lst)))
+
+;;; Core
 
 (defun most-used-words-buffer-1 (n &optional show-percentages-p)
   "Make a list of the N most used words in buffer.
 
-Optional argument SHOW-PERCENTAGES-P displays word counts and
-percentages."
-  (let ((counts (make-hash-table :test #'equal))
-	(total-count 0)
-	sorted-counts
-	start
-	end)
-    (save-excursion
-      (goto-char (point-min))
-      (skip-syntax-forward "^w")
-      (setf start (point))
-      (cl-loop until (eobp)
-	 do
-	   (skip-syntax-forward "w")
-	   (setf end (point))
-	   (cl-incf (gethash (buffer-substring start end) counts 0))
-	   (skip-syntax-forward "^w")
-	   (setf start (point))
-	   (cl-incf total-count)))
-    (cl-loop for word being the hash-keys of counts
-       using (hash-values count)
-       do
-	 (push (list word count) sorted-counts)
-       finally (setf sorted-counts (cl-sort sorted-counts #'>
-					  :key #'cl-second)))
-    (if show-percentages-p
-	(list (cl-subseq sorted-counts 0 n) total-count)
-      (mapcar #'cl-first (cl-subseq sorted-counts 0 n)))))
+Optional argument SHOW-PERCENTAGES-P displays word counts and percentages."
+  (with-current-buffer most-used-words--buffer
```

Comment:
This could be a bad solution, the better and comment solution is to add `&optional buffer-or-buffer-name` as an argument, so you can pass in buffer whatever you want! Also good for external use case!

Anyway, I assume this is good enough for now.

Reply:
I can't see any obvious problem at a first look. Again, I'm trusting your work.